### ISSUE TITLE
1-4 scaling not applied if lj14scale=0.0

### ISSUE THREAD

**Issue by rsdefever** [mosdef-hub/foyer#305]
**Bug summary**

If `lj14scale=0.0` the `structure.adjusts` section of the parameterized `parmed.structure` is empty. I would have expected the relevant 1-4 interactions to be listed. I may also be misunderstanding the expected behavior here.

**Code to reproduce the behavior**

Consider a united atom n-butane with TRAPPE.

```
import mbuild
import foyer

# Build the molecule
compound = mbuild.Compound(name='n-butane') 
c1 = mbuild.Compound(name='_CH3',pos=[0.,0.,0.]) 
c2 = mbuild.Compound(name='_CH2',pos=c1.pos+[0.154,0.,0.]) 
c3 = mbuild.Compound(name='_CH2',pos=c2.pos+[0.154,0.,0.]) 
c4 = mbuild.Compound(name='_CH3',pos=c3.pos+[0.154,0.,0.]) 
compound.add(c1) 
compound.add(c2) 
compound.add(c3) 
compound.add(c4) 
compound.add_bond([c1,c2]) 
compound.add_bond([c2,c3]) 
compound.add_bond([c3,c4]) 

# Load TRAPPE
trappe = foyer.forcefields.load_TRAPPE_UA()

# Apply FF
structure = trappe.apply(compound, combining_rule='lorentz')

# Check structure.adjusts
structure.adjusts 
Out : TrackedList([
])
```

If the TRAPPE xml file is edited so that the lj14scale=0.1 (rather than 0.0), then `structure.adjusts` is populated by: `TrackedList([<NonbondedException; <Atom _CH3 [0]; In RES 0> and <Atom _CH3 [3]; In RES 0>, type=<NonbondedExceptionType; rmin=4.2092, epsilon=0.0195, chgscale=1.0000>>
])`

Note that since there are no charges in the system the `chgscale=1.0` in the `NonbondedException` regardless of the selected `chgscale` value in the `xml` file.

**Comment by mattwthompson** [comment]
I'm not sure how this is implemented in other engines, but in GROMACS, the option `nrexcl` option set to 3 means the 1-4 interactions are ignored, and this factor being 0.0 means they're not added back. (The typical behavior, say using a factor of 0.5, would result in think these interactions excluded and then "added" back, scaled by 0.5). So, in GROMACS world, this seems like the right behavior to me. Is the issue related to its implementation in another engine? I believe Jason has some routined in ParmEd that remove parameters with no interaction, which would be responsible for this. If you need to carry "zero-interaction" types of things, it may be hackable from there.

**Comment by ahy3nz** [comment]
Get ready for a long-winded explanation:

* When we denote `coulomb14scale` or `lj14scale` in an FFXML, it gets used when `foyer` calls `createSystem` 
* in `foyer.Forcefield.createSystem`, we iterate through all the `openmm.generators` to call `createForce` and `postprocessSystem`
* in `openmm.NonbondedGenerator`, we call `openmm.NonbondedForce.createExceptionsFromBond` [here](https://github.com/openmm/openmm/blob/b6dfe9ffd2d2ef0bbe949292866c61532b058d3e/wrappers/python/simtk/openmm/app/forcefield.py#L2402), which is the OpenMM way of modifying 14 interactions, by passing information about `NonbondedGenerator.coulomb14scale` and `NonbondedGenerator.lj14scale` (which correspond to the FFXML)
* All this OpenMM work has happened [in this foyer line](https://github.com/mosdef-hub/foyer/blob/5790b068907b9aec38fe2964937aaabca82e3a48/foyer/forcefield.py#L595)
* However, OpenMM -> Parmed has lost some information [here](https://github.com/mosdef-hub/foyer/blob/5790b068907b9aec38fe2964937aaabca82e3a48/foyer/forcefield.py#L601)
* More specifically, when parmed tries to load the openmm system, parmed tries to process exceptions [here](https://github.com/ParmEd/ParmEd/blob/8fd6f492d352aaa3a71327d6e7a10bd8a74b6a4b/parmed/openmm/topsystem.py#L419), and I think it's somewhere in `parmed/openmm/topsystem.py` where 14 exclusions (neglecting any and all 14 LJ and QQ interactions) are lost and don't translate well into the parmed structure

So here are some workarounds
1) `my_foyerforcefield_object.getGenerators()[3].coulomb14scale` and `my_foyerforcefield_object.getGenerators()[3].lj14scale` will tell you exactly what the 14scalings in your FFXML are. You can parse that information from your Forcefield object and do whatever you want with it
2) Fix the parmed logic with processing `coulomb14scale=0.0` and `lj14scale=0.0` 

But yes if gmx fudgeLJ=1, then you're multiplying 14 LJ interactions by 1.0, if gmx fudgeLJ=0, then you're multiplying 14 LJ interactions by 0.0. This is similarly reflected in the FFXMLs also in `coulomb14scale` and `lj14scale`

**Comment by rsdefever** [comment]
Thanks for the feedback! I need to more closely consider several points in both of your answers, but in the meantime I'll provide some additional context. 

In Cassandra we provide the 1-4 scaling parameters (e.g., a number between 0.0 and 1.0) for LJ and charge in the MCF file. My goal is to write an MCF file given just a `parmed.Structure` object for a single compound. The original implementation infers the 1-4 LJ scaling parameter from a comparison of `epsilon` in `structure.adjusts` and `structure.atoms`. I am open to better approaches! 

Thanks again for the feedback. I'll give this a closer look tomorrow.

**Comment by mattwthompson** [comment]
> The original implementation infers the 1-4 LJ scaling parameter from a comparison of epsilon in structure.adjusts and structure.atoms.

This may not be the most elegant approach but it's what Jason did in ParmEd's GROMACS TOP writer (which seems to be doing the same thing as you need for Cassandra, taking in a bunch of adjusts and extracting a single scaling factor) and what we did in foyer's XML writer. I'm not sure the best way to do this using ParmEd, our backend may be able to do this cleaner (or there may be other issues down the line, it's hard to say)

https://github.com/ParmEd/ParmEd/blob/8fd6f492d352aaa3a71327d6e7a10bd8a74b6a4b/parmed/gromacs/gromacstop.py#L1265-L1281

https://github.com/mosdef-hub/foyer/blob/5790b068907b9aec38fe2964937aaabca82e3a48/foyer/xml_writer.py#L355-L399

**Comment by ahy3nz** [comment]
Parmed's approach to derive 14scaling was to look at the `scee` and `scnb` values of each DihedralType, but it doesn't look like parmed derives scee/scnb values when loading in the openmm system. So, I can't think of a situation where our foyer XML leads to populating the scee/scnb properties 

The foyer xml writer derives 14scaling from the `structure.adjusts` attribute, but the original issue is that `structure.adjusts` never gets populated if the XML says `coulomb14scale=0.0` and `lj14scale=0.0`

If `len(structure.adjusts)==0`, I wonder if you can 100% infer with certainty that 1-4 interactions should be completely excluded

**Comment by daico007** [comment]
Should be fixed by #496
